### PR TITLE
Overwrite xunit's app domain handling to not call AppDomain.Unload

### LINKED ISSUE #15497: "Attempted to access an unloaded AppDomain" during Roslyn.Services.Editor2.UnitTests.dll
At the end of the xunit log:

  Finished:    Roslyn.Services.Editor2.UnitTests
System.AppDomainUnloadedException: Attempted to access an unloaded AppDomain.

Unfortunately no additional output or crash dump was produced.

### LINKED ISSUE #40369: Unit test deadlock during xUnit Shutdown
This was hit in a CI run. The unit test shutdown of the AppDomain hung here:

```
 	clr.dll!MsgWaitHelper(int numWaiters, void * * phEvent, int bWaitAll, unsigned long millis, int bAlertable) Line 4258	C++
 	[Inline Frame] clr.dll!Thread::DoAppropriateAptStateWait(int) Line 4323	C++
 	clr.dll!Thread::DoAppropriateWaitWorker(int countHandles, void * * handles, int waitAll, unsigned long millis, WaitMode mode) Line 4468	C++
 	clr.dll!Thread::DoAppropriateWait(int countHandles, void * * handles, int waitAll, unsigned long millis, WaitMode mode, PendingSync * syncState) Line 4135	C++
 	[Inline Frame] clr.dll!SOIntolerantTransitionHandler::SetNoException() Line 853	C++
 	clr.dll!CLREventBase::WaitEx(unsigned long dwMilliseconds, WaitMode mode, PendingSync * syncState) Line 849	C++
 	clr.dll!AppDomain::UnloadWait(ADID Id, ADUnloadSink * pSink) Line 11187	C++
 	clr.dll!AppDomain::UnloadById(ADID dwId, int fSync, int) Line 11166	C++
 	clr.dll!AppDomainNative::Unload(int dwId) Line 1235	C++
```

### LINKED ISSUE #48989: Microsoft.VisualStudio.LanguageServices.CSharp.UnitTests is timing out
This test is timing out in our unit test runs. 

|Build|Kind|Start Time|
|---|---|---|
[865229](https://dev.azure.com/dnceng/public/_build/results?buildId=865229)|Rolling|2020-25-10|
[868161](https://dev.azure.com/dnceng/public/_build/results?buildId=868161)|Rolling|2020-27-10|
[867599](https://dev.azure.com/dnceng/public/_build/results?buildId=867599)|Rolling|2020-27-10|
[866169](https://dev.azure.com/dnceng/public/_build/results?buildId=866169)|Rolling|2020-26-10|
[865435](https://dev.azure.com/dnceng/public/_build/results?buildId=865435)|Rolling|2020-25-10|
[868290](https://dev.azure.com/dnceng/public/_build/results?buildId=868290)|Rolling|2020-28-10|

### LINKED ISSUE #49024: Test timeout causes CI to fail
This is a tracking issue for builds that fail due to test timeout, but were not further identified to have a specific root cause.

|Build|Kind|Start Time|
|---|---|---|
[867006](https://dev.azure.com/dnceng/public/_build/results?buildId=867006)|[PR 48942](https://github.com/dotnet/roslyn/pull/48942)|2020-27-10|
[866756](https://dev.azure.com/dnceng/public/_build/results?buildId=866756)|[PR 48468](https://github.com/dotnet/roslyn/pull/48468)|2020-27-10|
[866527](https://dev.azure.com/dnceng/public/_build/results?buildId=866527)|[PR 47252](https://github.com/dotnet/roslyn/pull/47252)|2020-26-10|
[866497](https://dev.azure.com/dnceng/public/_build/results?buildId=866497)|Rolling|2020-26-10|
[866377](https://dev.azure.com/dnceng/public/_build/results?buildId=866377)|[PR 48858](https://github.com/dotnet/roslyn/pull/48858)|2020-26-10|
[866347](https://dev.azure.com/dnceng/public/_build/results?buildId=866347)|[PR 48215](https://github.com/dotnet/roslyn/pull/48215)|2020-26-10|
[865923](https://dev.azure.com/dnceng/public/_build/results?buildId=865923)|[PR 48896](https://github.com/dotnet/roslyn/pull/48896)|2020-26-10|
[865637](https://dev.azure.com/dnceng/public/_build/results?buildId=865637)|[PR 48915](https://github.com/dotnet/roslyn/pull/48915)|2020-26-10|
[865482](https://dev.azure.com/dnceng/public/_build/results?buildId=865482)|Rolling|2020-26-10|
[869857](https://dev.azure.com/dnceng/public/_build/results?buildId=869857)|[PR 48862](https://github.com/dotnet/roslyn/pull/48862)|2020-29-10|
[869589](https://dev.azure.com/dnceng/public/_build/results?buildId=869589)|[PR 49003](https://github.com/dotnet/roslyn/pull/49003)|2020-29-10|
[869220](https://dev.azure.com/dnceng/public/_build/results?buildId=869220)|[PR 48939](https://github.com/dotnet/roslyn/pull/48939)|2020-28-10|
[868941](https://dev.azure.com/dnceng/public/_build/results?buildId=868941)|[PR 48980](https://github.com/dotnet/roslyn/pull/48980)|2020-28-10|
[868353](https://dev.azure.com/dnceng/public/_build/results?buildId=868353)|[PR 48961](https://github.com/dotnet/roslyn/pull/48961)|2020-28-10|
[868281](https://dev.azure.com/dnceng/public/_build/results?buildId=868281)|[PR 48965](https://github.com/dotnet/roslyn/pull/48965)|2020-28-10|
[866785](https://dev.azure.com/dnceng/public/_build/results?buildId=866785)|[PR 47252](https://github.com/dotnet/roslyn/pull/47252)|2020-27-10|
[869487](https://dev.azure.com/dnceng/public/_build/results?buildId=869487)|Rolling|2020-28-10|
[869786](https://dev.azure.com/dnceng/public/_build/results?buildId=869786)|Rolling|2020-29-10|
[869068](https://dev.azure.com/dnceng/public/_build/results?buildId=869068)|Rolling|2020-28-10|
[869299](https://dev.azure.com/dnceng/public/_build/results?buildId=869299)|Rolling|2020-28-10|
[869442](https://dev.azure.com/dnceng/public/_build/results?buildId=869442)|Rolling|2020-28-10|
[869473](https://dev.azure.com/dnceng/public/_build/results?buildId=869473)|Rolling|2020-28-10|
[870401](https://dev.azure.com/dnceng/public/_build/results?buildId=870401)|[PR 49051](https://github.com/dotnet/roslyn/pull/49051)|2020-29-10|
[869534](https://dev.azure.com/dnceng/public/_build/results?buildId=869534)|Rolling|2020-28-10|
[870916](https://dev.azure.com/dnceng/public/_build/results?buildId=870916)|[PR 48967](https://github.com/dotnet/roslyn/pull/48967)|2020-30-10|
[870959](https://dev.azure.com/dnceng/public/_build/results?buildId=870959)|Rolling|2020-30-10|
[870873](https://dev.azure.com/dnceng/public/_build/results?buildId=870873)|Rolling|2020-30-10|
[870670](https://dev.azure.com/dnceng/public/_build/results?buildId=870670)|Rolling|2020-29-10|

### LINKED ISSUE #49035: Hang in AppDomain.Unload causes tests to time out
This is a tracking issue for failed CI jobs due to a hang in `AppDomain.Unload`:

```
 	[Managed to Native Transition]		Annotated Frame
 	mscorlib.dll!System.AppDomain.Unload(System.AppDomain domain) Line 2590	C#	Symbols loaded.
>	xunit.runner.utility.net452.dll!Xunit.AppDomainManager_AppDomain.Dispose() Line 99	C#	Symbols loaded.
 	xunit.runner.utility.net452.dll!Xunit.Xunit2Discoverer.Dispose() Line 164	C#	Symbols loaded.
 	xunit.runner.utility.net452.dll!Xunit.Xunit2.Dispose() Line 89	C#	Symbols loaded.
 	xunit.runner.utility.net452.dll!Xunit.XunitFrontController.Dispose() Line 155	C#	Symbols loaded.
 	xunit.runner.visualstudio.testadapter.dll!Xunit.Runner.VisualStudio.VsTestRunner.RunTestsInAssembly(Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IRunContext runContext, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle, LoggerHelper logger, Xunit.Runner.VisualStudio.TestPlatformContext testPlatformContext, Xunit.Runner.VisualStudio.RunSettings runSettings, Xunit.IMessageSinkWithTypes reporterMessageHandler, Xunit.Runner.VisualStudio.AssemblyRunInfo runInfo) Line 619	C#	Symbols loaded.
 	xunit.runner.visualstudio.testadapter.dll!Xunit.Runner.VisualStudio.VsTestRunner.RunTests.AnonymousMethod__4(Xunit.Runner.VisualStudio.AssemblyRunInfo runInfo) Line 452	C#	Symbols loaded.
 	mscorlib.dll!System.Collections.Generic.List<Xunit.Runner.VisualStudio.AssemblyRunInfo>.ForEach(System.Action<Xunit.Runner.VisualStudio.AssemblyRunInfo> action) Line 553	C#	Symbols loaded.
 	xunit.runner.visualstudio.testadapter.dll!Xunit.Runner.VisualStudio.VsTestRunner.RunTests(Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IRunContext runContext, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle, LoggerHelper logger, Xunit.Runner.VisualStudio.TestPlatformContext testPlatformContext, Xunit.Runner.VisualStudio.RunSettings runSettings, System.Func<System.Collections.Generic.List<Xunit.Runner.VisualStudio.AssemblyRunInfo>> getRunInfos) Line 451	C#	Symbols loaded.
 	xunit.runner.visualstudio.testadapter.dll!Xunit.Runner.VisualStudio.VsTestRunner.Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.ITestExecutor.RunTests(System.Collections.Generic.IEnumerable<string> sources, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IRunContext runContext, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle) Line 181	C#	Symbols loaded.
 	Microsoft.TestPlatform.CrossPlatEngine.dll!Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Execution.RunTestsWithSources.InvokeExecutor(Microsoft.VisualStudio.TestPlatform.Common.ExtensionFramework.Utilities.LazyExtension<Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.ITestExecutor, Microsoft.VisualStudio.TestPlatform.Common.Interfaces.ITestExecutorCapabilities> executor, System.Tuple<System.Uri, string> executorUriExtensionTuple, Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Adapter.RunContext runContext, Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.IFrameworkHandle frameworkHandle) Line 121	C#	Symbols loaded.
 	Microsoft.TestPlatform.CrossPlatEngine.dll!Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Execution.BaseRunTests.RunTestInternalWithExecutors.AnonymousMethod__0() Line 494	C#	Symbols loaded.
 	Microsoft.TestPlatform.PlatformAbstractions.dll!Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.PlatformThread.Run.AnonymousMethod__0() Line 26	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ThreadHelper.ThreadStart_Context(object state) Line 74	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 954	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 902	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 891	C#	Symbols loaded.
 	mscorlib.dll!System.Threading.ThreadHelper.ThreadStart() Line 111	C#	Symbols loaded.
```

### LINKED ISSUE #49275: Roslyn unit tests failing during xUnit AppDomain unload combined with vs-threading FailFast
When Roslyn unit tests shut down, xUnit unloads the AppDomain that our tests are loaded into. This means that the test threads inside the AppDomain are unloaded, which is done by aborting the threads. This triggers this FailFast here:

https://github.com/microsoft/vs-threading/blob/a2c873656dd7a7967499595a4587cee7b71ff4b4/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs#L968-L974

which causes the entire test process to crash.

This is a tracking bug so I can link failures in other PRs to it as I find them.

### PR DESCRIPTION
**Pull request by sharwell**
Hoping this avoids the need to maintain our own copy of the xunit runners, by patching the implementation at runtime to apply https://github.com/xunit/xunit/pull/2192.

Fixes #15497
Fixes #40369
Fixes #48989
Fixes #49024
Fixes #49035
Fixes #49275

### CODE REVIEW

File: src/EditorFeatures/XunitHook/XunitDisposeHook.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                    // lea esp,[ebp-8]
+                    Marshal.WriteByte(functionPointer, offset++, 0x8D);
+                    Marshal.WriteByte(functionPointer, offset++, 0x65);
+                    Marshal.WriteByte(functionPointer, offset++, 0xF8);
+
+                    // pop ebp
+                    Marshal.WriteByte(functionPointer, offset++, 0x5D);
+
+                    // ret
+                    Marshal.WriteByte(functionPointer, offset++, 0xC3);
```
Comment (sharwell) [inline src/EditorFeatures/XunitHook/XunitDisposeHook.cs:53]: 💡 This is overly complicated. Only the final `ret` instruction is needed for this case.

File: src/EditorFeatures/XunitHook/XunitDisposeHook.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+                    // ret
+                    Marshal.WriteByte(functionPointer, 0xC3);
+                }
+                else
+                {
+                    // Overwrite the compiled method to just return
+
+                    // ret
+                    Marshal.WriteByte(functionPointer, 0xC3);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/XunitHook/XunitDisposeHook.cs:43]: lalalala, i'm not seeing anything.

**Review by CyrusNajmabadi — APPROVED** [review]
fine with this as long as we have a comment linking to wahtever issue/pr we raised on the xunit side that we couldn't effectivel move forward with.

File: src/EditorFeatures/XunitHook/XunitDisposeHook.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+            foreach (var xunitUtility in xunitUtilities)
+            {
+                var appDomainManagerType = xunitUtility.GetType("Xunit.AppDomainManager_AppDomain");
+                if (appDomainManagerType is null)
+                    continue;
+
+                var method = appDomainManagerType.GetMethod("Dispose");
+                RuntimeHelpers.PrepareMethod(method.MethodHandle);
+                var functionPointer = method.MethodHandle.GetFunctionPointer();
+                if (IntPtr.Size == 4)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/XunitHook/XunitDisposeHook.cs:31]: ➡️ my next push consolidates these with a note that both x86 and x64 use the same code

File: src/EditorFeatures/XunitHook/XunitDisposeHook.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+                {
+                    // Overwrite the compiled method to just return
+
+                    // ret
+                    Marshal.WriteByte(functionPointer, 0xC3);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/XunitHook/XunitDisposeHook.cs:36]: ➡️ my next push checks `RuntimeInformation.ProcessArchitecture`

File: src/EditorFeatures/XunitHook/XunitDisposeHook.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/XunitHook/XunitDisposeHook.cs]: Now pushed

File: src/EditorFeatures/XunitHook/XunitDisposeHook.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/EditorFeatures/XunitHook/XunitDisposeHook.cs]: Now pushed

**Comment by jaredpar** [comment]
This is beautiful

**Pull request merged**